// vue3-fortune-wheel/dist/wheel-lib.es.js
import { defineComponent as se, ref as z, computed as gn, onMounted as ue, openBlock as le, createElementBlock as fe, normalizeClass as ce, normalizeStyle as he, unref as de } from "vue";
var pe = { value: () => {
} };
function Fn() {
  for (var t = 0, n = arguments.length, e = {}, r; t < n; ++t) {
    if (!(r = arguments[t] + "") || r in e || /[\s.]/.test(r))
      throw new Error("illegal type: " + r);
    e[r] = [];
  }
  return new Nt(e);
}
function Nt(t) {
  this._ = t;
}
function ge(t, n) {
  return t.trim().split(/^|\s+/).map(function(e) {
    var r = "", i = e.indexOf(".");
    if (i >= 0 && (r = e.slice(i + 1), e = e.slice(0, i)), e && !n.hasOwnProperty(e))
      throw new Error("unknown type: " + e);
    return { type: e, name: r };
  });
}
Nt.prototype = Fn.prototype = {
  constructor: Nt,
  on: function(t, n) {
    var e = this._, r = ge(t + "", e), i, o = -1, a = r.length;
    if (arguments.length < 2) {
      for (; ++o < a; )
        if ((i = (t = r[o]).type) && (i = ye(e[i], t.name)))
          return i;
      return;
    }
    if (n != null && typeof n != "function")
      throw new Error("invalid callback: " + n);
    for (; ++o < a; )
      if (i = (t = r[o]).type)
        e[i] = yn(e[i], t.name, n);
      else if (n == null)
        for (i in e)
          e[i] = yn(e[i], t.name, null);
    return this;
  },
  copy: function() {
    var t = {}, n = this._;
    for (var e in n)
      t[e] = n[e].slice();
    return new Nt(t);
  },
  call: function(t, n) {
    if ((i = arguments.length - 2) > 0)
      for (var e = new Array(i), r = 0, i, o; r < i; ++r)
        e[r] = arguments[r + 2];
    if (!this._.hasOwnProperty(t))
      throw new Error("unknown type: " + t);
    for (o = this._[t], r = 0, i = o.length; r < i; ++r)
      o[r].value.apply(n, e);
  },
  apply: function(t, n, e) {
    if (!this._.hasOwnProperty(t))
      throw new Error("unknown type: " + t);
    for (var r = this._[t], i = 0, o = r.length; i < o; ++i)
      r[i].value.apply(n, e);
  }
};
function ye(t, n) {
  for (var e = 0, r = t.length, i; e < r; ++e)
    if ((i = t[e]).name === n)
      return i.value;
}
function yn(t, n, e) {
  for (var r = 0, i = t.length; r < i; ++r)
    if (t[r].name === n) {
      t[r] = pe, t = t.slice(0, r).concat(t.slice(r + 1));
      break;
    }
  return e != null && t.push({ name: n, value: e }), t;
}
var Gt = "http://www.w3.org/1999/xhtml";
const _n = {
  svg: "http://www.w3.org/2000/svg",
  xhtml: Gt,
  xlink: "http://www.w3.org/1999/xlink",
  xml: "http://www.w3.org/XML/1998/namespace",
  xmlns: "http://www.w3.org/2000/xmlns/"
};
function Dt(t) {
  var n = t += "", e = n.indexOf(":");
  return e >= 0 && (n = t.slice(0, e)) !== "xmlns" && (t = t.slice(e + 1)), _n.hasOwnProperty(n) ? { space: _n[n], local: t } : t;
}
function _e(t) {
  return function() {
    var n = this.ownerDocument, e = this.namespaceURI;
    return e === Gt && n.documentElement.namespaceURI === Gt ? n.createElement(t) : n.createElementNS(e, t);
  };
}
function ve(t) {
  return function() {
    return this.ownerDocument.createElementNS(t.space, t.local);
  };
}
function Pn(t) {
  var n = Dt(t);
  return (n.local ? ve : _e)(n);
}
function me() {
}
function an(t) {
  return t == null ? me : function() {
    return this.querySelector(t);
  };
}
function xe(t) {
  typeof t != "function" && (t = an(t));
  for (var n = this._groups, e = n.length, r = new Array(e), i = 0; i < e; ++i)
    for (var o = n[i], a = o.length, s = r[i] = new Array(a), u, l, f = 0; f < a; ++f)
      (u = o[f]) && (l = t.call(u, u.__data__, f, o)) && ("__data__" in u && (l.__data__ = u.__data__), s[f] = l);
  return new P(r, this._parents);
}
function we(t) {
  return t == null ? [] : Array.isArray(t) ? t : Array.from(t);
}
function Ae() {
  return [];
}
function On(t) {
  return t == null ? Ae : function() {
    return this.querySelectorAll(t);
  };
}
function be(t) {
  return function() {
    return we(t.apply(this, arguments));
  };
}
function $e(t) {
  typeof t == "function" ? t = be(t) : t = On(t);
  for (var n = this._groups, e = n.length, r = [], i = [], o = 0; o < e; ++o)
    for (var a = n[o], s = a.length, u, l = 0; l < s; ++l)
      (u = a[l]) && (r.push(t.call(u, u.__data__, l, a)), i.push(u));
  return new P(r, i);
}
function Vn(t) {
  return function() {
    return this.matches(t);
  };
}
function Dn(t) {
  return function(n) {
    return n.matches(t);
  };
}
var Ne = Array.prototype.find;
function Me(t) {
  return function() {
    return Ne.call(this.children, t);
  };
}
function ke() {
  return this.firstElementChild;
}
function Se(t) {
  return this.select(t == null ? ke : Me(typeof t == "function" ? t : Dn(t)));
}
var Ee = Array.prototype.filter;
function Te() {
  return Array.from(this.children);
}
function Ce(t) {
  return function() {
    return Ee.call(this.children, t);
  };
}
function Re(t) {
  return this.selectAll(t == null ? Te : Ce(typeof t == "function" ? t : Dn(t)));
}
function Ie(t) {
  typeof t != "function" && (t = Vn(t));
  for (var n = this._groups, e = n.length, r = new Array(e), i = 0; i < e; ++i)
    for (var o = n[i], a = o.length, s = r[i] = [], u, l = 0; l < a; ++l)
      (u = o[l]) && t.call(u, u.__data__, l, o) && s.push(u);
  return new P(r, this._parents);
}
function qn(t) {
  return new Array(t.length);
}
function Fe() {
  return new P(this._enter || this._groups.map(qn), this._parents);
}
function Et(t, n) {
  this.ownerDocument = t.ownerDocument, this.namespaceURI = t.namespaceURI, this._next = null, this._parent = t, this.__data__ = n;
}
Et.prototype = {
  constructor: Et,
  appendChild: function(t) {
    return this._parent.insertBefore(t, this._next);
  },
  insertBefore: function(t, n) {
    return this._parent.insertBefore(t, n);
  },
  querySelector: function(t) {
    return this._parent.querySelector(t);
  },
  querySelectorAll: function(t) {
    return this._parent.querySelectorAll(t);
  }
};
function Pe(t) {
  return function() {
    return t;
  };
}
function Oe(t, n, e, r, i, o) {
  for (var a = 0, s, u = n.length, l = o.length; a < l; ++a)
    (s = n[a]) ? (s.__data__ = o[a], r[a] = s) : e[a] = new Et(t, o[a]);
  for (; a < u; ++a)
    (s = n[a]) && (i[a] = s);
}
function Ve(t, n, e, r, i, o, a) {
  var s, u, l = /* @__PURE__ */ new Map(), f = n.length, c = o.length, h = new Array(f), d;
  for (s = 0; s < f; ++s)
    (u = n[s]) && (h[s] = d = a.call(u, u.__data__, s, n) + "", l.has(d) ? i[s] = u : l.set(d, u));
  for (s = 0; s < c; ++s)
    d = a.call(t, o[s], s, o) + "", (u = l.get(d)) ? (r[s] = u, u.__data__ = o[s], l.delete(d)) : e[s] = new Et(t, o[s]);
  for (s = 0; s < f; ++s)
    (u = n[s]) && l.get(h[s]) === u && (i[s] = u);
}
function De(t) {
  return t.__data__;
}
function qe(t, n) {
  if (!arguments.length)
    return Array.from(this, De);
  var e = n ? Ve : Oe, r = this._parents, i = this._groups;
  typeof t != "function" && (t = Pe(t));
  for (var o = i.length, a = new Array(o), s = new Array(o), u = new Array(o), l = 0; l < o; ++l) {
    var f = r[l], c = i[l], h = c.length, d = Le(t.call(f, f && f.__data__, l, r)), p = d.length, y = s[l] = new Array(p), _ = a[l] = new Array(p), N = u[l] = new Array(h);
    e(f, c, y, _, N, d, n);
    for (var w = 0, x = 0, A, M; w < p; ++w)
      if (A = y[w]) {
        for (w >= x && (x = w + 1); !(M = _[x]) && ++x < p; )
          ;
        A._next = M || null;
      }
  }
  return a = new P(a, r), a._enter = s, a._exit = u, a;
}
function Le(t) {
  return typeof t == "object" && "length" in t ? t : Array.from(t);
}
function He() {
  return new P(this._exit || this._groups.map(qn), this._parents);
}
function Xe(t, n, e) {
  var r = this.enter(), i = this, o = this.exit();
  return typeof t == "function" ? (r = t(r), r && (r = r.selection())) : r = r.append(t + ""), n != null && (i = n(i), i && (i = i.selection())), e == null ? o.remove() : e(o), r && i ? r.merge(i).order() : i;
}
function ze(t) {
  for (var n = t.selection ? t.selection() : t, e = this._groups, r = n._groups, i = e.length, o = r.length, a = Math.min(i, o), s = new Array(i), u = 0; u < a; ++u)
    for (var l = e[u], f = r[u], c = l.length, h = s[u] = new Array(c), d, p = 0; p < c; ++p)
      (d = l[p] || f[p]) && (h[p] = d);
  for (; u < i; ++u)
    s[u] = e[u];
  return new P(s, this._parents);
}
function Be() {
  for (var t = this._groups, n = -1, e = t.length; ++n < e; )
    for (var r = t[n], i = r.length - 1, o = r[i], a; --i >= 0; )
      (a = r[i]) && (o && a.compareDocumentPosition(o) ^ 4 && o.parentNode.insertBefore(a, o), o = a);
  return this;
}
function Ye(t) {
  t || (t = We);
  function n(c, h) {
    return c && h ? t(c.__data__, h.__data__) : !c - !h;
  }
  for (var e = this._groups, r = e.length, i = new Array(r), o = 0; o < r; ++o) {
    for (var a = e[o], s = a.length, u = i[o] = new Array(s), l, f = 0; f < s; ++f)
      (l = a[f]) && (u[f] = l);
    u.sort(n);
  }
  return new P(i, this._parents).order();
}
function We(t, n) {
  return t < n ? -1 : t > n ? 1 : t >= n ? 0 : NaN;
}
function Ue() {
  var t = arguments[0];
  return arguments[0] = this, t.apply(null, arguments), this;
}
function Ge() {
  return Array.from(this);
}
function Ke() {
  for (var t = this._groups, n = 0, e = t.length; n < e; ++n)
    for (var r = t[n], i = 0, o = r.length; i < o; ++i) {
      var a = r[i];
      if (a)
        return a;
    }
  return null;
}
function Ze() {
  let t = 0;
  for (const n of this)
    ++t;
  return t;
}
function Qe() {
  return !this.node();
}
function Je(t) {
  for (var n = this._groups, e = 0, r = n.length; e < r; ++e)
    for (var i = n[e], o = 0, a = i.length, s; o < a; ++o)
      (s = i[o]) && t.call(s, s.__data__, o, i);
  return this;
}
function je(t) {
  return function() {
    this.removeAttribute(t);
  };
}
function tr(t) {
  return function() {
    this.removeAttributeNS(t.space, t.local);
  };
}
function nr(t, n) {
  return function() {
    this.setAttribute(t, n);
  };
}
function er(t, n) {
  return function() {
    this.setAttributeNS(t.space, t.local, n);
  };
}
function rr(t, n) {
  return function() {
    var e = n.apply(this, arguments);
    e == null ? this.removeAttribute(t) : this.setAttribute(t, e);
  };
}
function ir(t, n) {
  return function() {
    var e = n.apply(this, arguments);
    e == null ? this.removeAttributeNS(t.space, t.local) : this.setAttributeNS(t.space, t.local, e);
  };
}
function ar(t, n) {
  var e = Dt(t);
  if (arguments.length < 2) {
    var r = this.node();
    return e.local ? r.getAttributeNS(e.space, e.local) : r.getAttribute(e);
  }
  return this.each((n == null ? e.local ? tr : je : typeof n == "function" ? e.local ? ir : rr : e.local ? er : nr)(e, n));
}
function Ln(t) {
  return t.ownerDocument && t.ownerDocument.defaultView || t.document && t || t.defaultView;
}
function or(t) {
  return function() {
    this.style.removeProperty(t);
  };
}
function sr(t, n, e) {
  return function() {
    this.style.setProperty(t, n, e);
  };
}
function ur(t, n, e) {
  return function() {
    var r = n.apply(this, arguments);
    r == null ? this.style.removeProperty(t) : this.style.setProperty(t, r, e);
  };
}
function lr(t, n, e) {
  return arguments.length > 1 ? this.each((n == null ? or : typeof n == "function" ? ur : sr)(t, n, e ?? "")) : lt(this.node(), t);
}
function lt(t, n) {
  return t.style.getPropertyValue(n) || Ln(t).getComputedStyle(t, null).getPropertyValue(n);
}
function fr(t) {
  return function() {
    delete this[t];
  };
}
function cr(t, n) {
  return function() {
    this[t] = n;
  };
}
function hr(t, n) {
  return function() {
    var e = n.apply(this, arguments);
    e == null ? delete this[t] : this[t] = e;
  };
}
function dr(t, n) {
  return arguments.length > 1 ? this.each((n == null ? fr : typeof n == "function" ? hr : cr)(t, n)) : this.node()[t];
}
function Hn(t) {
  return t.trim().split(/^|\s+/);
}
function on(t) {
  return t.classList || new Xn(t);
}
function Xn(t) {
  this._node = t, this._names = Hn(t.getAttribute("class") || "");
}
Xn.prototype = {
  add: function(t) {
    var n = this._names.indexOf(t);
    n < 0 && (this._names.push(t), this._node.setAttribute("class", this._names.join(" ")));
  },
  remove: function(t) {
    var n = this._names.indexOf(t);
    n >= 0 && (this._names.splice(n, 1), this._node.setAttribute("class", this._names.join(" ")));
  },
  contains: function(t) {
    return this._names.indexOf(t) >= 0;
  }
};
function zn(t, n) {
  for (var e = on(t), r = -1, i = n.length; ++r < i; )
    e.add(n[r]);
}
function Bn(t, n) {
  for (var e = on(t), r = -1, i = n.length; ++r < i; )
    e.remove(n[r]);
}
function pr(t) {
  return function() {
    zn(this, t);
  };
}
function gr(t) {
  return function() {
    Bn(this, t);
  };
}
function yr(t, n) {
  return function() {
    (n.apply(this, arguments) ? zn : Bn)(this, t);
  };
}
function _r(t, n) {
  var e = Hn(t + "");
  if (arguments.length < 2) {
    for (var r = on(this.node()), i = -1, o = e.length; ++i < o; )
      if (!r.contains(e[i]))
        return !1;
    return !0;
  }
  return this.each((typeof n == "function" ? yr : n ? pr : gr)(e, n));
}
function vr() {
  this.textContent = "";
}
function mr(t) {
  return function() {
    this.textContent = t;
  };
}
function xr(t) {
  return function() {
    var n = t.apply(this, arguments);
    this.textContent = n ?? "";
  };
}
function wr(t) {
  return arguments.length ? this.each(t == null ? vr : (typeof t == "function" ? xr : mr)(t)) : this.node().textContent;
}
function Ar() {
  this.innerHTML = "";
}
function br(t) {
  return function() {
    this.innerHTML = t;
  };
}
function $r(t) {
  return function() {
    var n = t.apply(this, arguments);
    this.innerHTML = n ?? "";
  };
}
function Nr(t) {
  return arguments.length ? this.each(t == null ? Ar : (typeof t == "function" ? $r : br)(t)) : this.node().innerHTML;
}
function Mr() {
  this.nextSibling && this.parentNode.appendChild(this);
}
function kr() {
  return this.each(Mr);
}
function Sr() {
  this.previousSibling && this.parentNode.insertBefore(this, this.parentNode.firstChild);
}
function Er() {
  return this.each(Sr);
}
function Tr(t) {
  var n = typeof t == "function" ? t : Pn(t);
  return this.select(function() {
    return this.appendChild(n.apply(this, arguments));
  });
}
function Cr() {
  return null;
}
function Rr(t, n) {
  var e = typeof t == "function" ? t : Pn(t), r = n == null ? Cr : typeof n == "function" ? n : an(n);
  return this.select(function() {
    return this.insertBefore(e.apply(this, arguments), r.apply(this, arguments) || null);
  });
}
function Ir() {
  var t = this.parentNode;
  t && t.removeChild(this);
}
function Fr() {
  return this.each(Ir);
}
function Pr() {
  var t = this.cloneNode(!1), n = this.parentNode;
  return n ? n.insertBefore(t, this.nextSibling) : t;
}
function Or() {
  var t = this.cloneNode(!0), n = this.parentNode;
  return n ? n.insertBefore(t, this.nextSibling) : t;
}
function Vr(t) {
  return this.select(t ? Or : Pr);
}
function Dr(t) {
  return arguments.length ? this.property("__data__", t) : this.node().__data__;
}
function qr(t) {
  return function(n) {
    t.call(this, n, this.__data__);
  };
}
function Lr(t) {
  return t.trim().split(/^|\s+/).map(function(n) {
    var e = "", r = n.indexOf(".");
    return r >= 0 && (e = n.slice(r + 1), n = n.slice(0, r)), { type: n, name: e };
  });
}
function Hr(t) {
  return function() {
    var n = this.__on;
    if (!!n) {
      for (var e = 0, r = -1, i = n.length, o; e < i; ++e)
        o = n[e], (!t.type || o.type === t.type) && o.name === t.name ? this.removeEventListener(o.type, o.listener, o.options) : n[++r] = o;
      ++r ? n.length = r : delete this.__on;
    }
  };
}
function Xr(t, n, e) {
  return function() {
    var r = this.__on, i, o = qr(n);
    if (r) {
      for (var a = 0, s = r.length; a < s; ++a)
        if ((i = r[a]).type === t.type && i.name === t.name) {
          this.removeEventListener(i.type, i.listener, i.options), this.addEventListener(i.type, i.listener = o, i.options = e), i.value = n;
          return;
        }
    }
    this.addEventListener(t.type, o, e), i = { type: t.type, name: t.name, value: n, listener: o, options: e }, r ? r.push(i) : this.__on = [i];
  };
}
function zr(t, n, e) {
  var r = Lr(t + ""), i, o = r.length, a;
  if (arguments.length < 2) {
    var s = this.node().__on;
    if (s) {
      for (var u = 0, l = s.length, f; u < l; ++u)
        for (i = 0, f = s[u]; i < o; ++i)
          if ((a = r[i]).type === f.type && a.name === f.name)
            return f.value;
    }
    return;
  }
  for (s = n ? Xr : Hr, i = 0; i < o; ++i)
    this.each(s(r[i], n, e));
  return this;
}
function Yn(t, n, e) {
  var r = Ln(t), i = r.CustomEvent;
  typeof i == "function" ? i = new i(n, e) : (i = r.document.createEvent("Event"), e ? (i.initEvent(n, e.bubbles, e.cancelable), i.detail = e.detail) : i.initEvent(n, !1, !1)), t.dispatchEvent(i);
}
function Br(t, n) {
  return function() {
    return Yn(this, t, n);
  };
}
function Yr(t, n) {
  return function() {
    return Yn(this, t, n.apply(this, arguments));
  };
}
function Wr(t, n) {
  return this.each((typeof n == "function" ? Yr : Br)(t, n));
}
function* Ur() {
  for (var t = this._groups, n = 0, e = t.length; n < e; ++n)
    for (var r = t[n], i = 0, o = r.length, a; i < o; ++i)
      (a = r[i]) && (yield a);
}
var Wn = [null];
function P(t, n) {
  this._groups = t, this._parents = n;
}
function mt() {
  return new P([[document.documentElement]], Wn);
}
function Gr() {
  return this;
}
P.prototype = mt.prototype = {
  constructor: P,
  select: xe,
  selectAll: $e,
  selectChild: Se,
  selectChildren: Re,
  filter: Ie,
  data: qe,
  enter: Fe,
  exit: He,
  join: Xe,
  merge: ze,
  selection: Gr,
  order: Be,
  sort: Ye,
  call: Ue,
  nodes: Ge,
  node: Ke,
  size: Ze,
  empty: Qe,
  each: Je,
  attr: ar,
  style: lr,
  property: dr,
  classed: _r,
  text: wr,
  html: Nr,
  raise: kr,
  lower: Er,
  append: Tr,
  insert: Rr,
  remove: Fr,
  clone: Vr,
  datum: Dr,
  on: zr,
  dispatch: Wr,
  [Symbol.iterator]: Ur
};
function vn(t) {
  return typeof t == "string" ? new P([[document.querySelector(t)]], [document.documentElement]) : new P([[t]], Wn);
}
function sn(t, n, e) {
  t.prototype = n.prototype = e, e.constructor = t;
}
function Un(t, n) {
  var e = Object.create(t.prototype);
  for (var r in n)
    e[r] = n[r];
  return e;
}
function xt() {
}
var gt = 0.7, Tt = 1 / gt, ut = "\\s*([+-]?\\d+)\\s*", yt = "\\s*([+-]?(?:\\d*\\.)?\\d+(?:[eE][+-]?\\d+)?)\\s*", W = "\\s*([+-]?(?:\\d*\\.)?\\d+(?:[eE][+-]?\\d+)?)%\\s*", Kr = /^#([0-9a-f]{3,8})$/, Zr = new RegExp(`^rgb\\(${ut},${ut},${ut}\\)$`), Qr = new RegExp(`^rgb\\(${W},${W},${W}\\)$`), Jr = new RegExp(`^rgba\\(${ut},${ut},${ut},${yt}\\)$`), jr = new RegExp(`^rgba\\(${W},${W},${W},${yt}\\)$`), ti = new RegExp(`^hsl\\(${yt},${W},${W}\\)$`), ni = new RegExp(`^hsla\\(${yt},${W},${W},${yt}\\)$`), mn = {
  aliceblue: 15792383,
  antiquewhite: 16444375,
  aqua: 65535,
  aquamarine: 8388564,
  azure: 15794175,
  beige: 16119260,
  bisque: 16770244,
  black: 0,
  blanchedalmond: 16772045,
  blue: 255,
  blueviolet: 9055202,
  brown: 10824234,
  burlywood: 14596231,
  cadetblue: 6266528,
  chartreuse: 8388352,
  chocolate: 13789470,
  coral: 16744272,
  cornflowerblue: 6591981,
  cornsilk: 16775388,
  crimson: 14423100,
  cyan: 65535,
  darkblue: 139,
  darkcyan: 35723,
  darkgoldenrod: 12092939,
  darkgray: 11119017,
  darkgreen: 25600,
  darkgrey: 11119017,
  darkkhaki: 12433259,
  darkmagenta: 9109643,
  darkolivegreen: 5597999,
  darkorange: 16747520,
  darkorchid: 10040012,
  darkred: 9109504,
  darksalmon: 15308410,
  darkseagreen: 9419919,
  darkslateblue: 4734347,
  darkslategray: 3100495,
  darkslategrey: 3100495,
  darkturquoise: 52945,
  darkviolet: 9699539,
  deeppink: 16716947,
  deepskyblue: 49151,
  dimgray: 6908265,
  dimgrey: 6908265,
  dodgerblue: 2003199,
  firebrick: 11674146,
  floralwhite: 16775920,
  forestgreen: 2263842,
  fuchsia: 16711935,
  gainsboro: 14474460,
  ghostwhite: 16316671,
  gold: 16766720,
  goldenrod: 14329120,
  gray: 8421504,
  green: 32768,
  greenyellow: 11403055,
  grey: 8421504,
  honeydew: 15794160,
  hotpink: 16738740,
  indianred: 13458524,
  indigo: 4915330,
  ivory: 16777200,
  khaki: 15787660,
  lavender: 15132410,
  lavenderblush: 16773365,
  lawngreen: 8190976,
  lemonchiffon: 16775885,
  lightblue: 11393254,
  lightcoral: 15761536,
  lightcyan: 14745599,
  lightgoldenrodyellow: 16448210,
  lightgray: 13882323,
  lightgreen: 9498256,
  lightgrey: 13882323,
  lightpink: 16758465,
  lightsalmon: 16752762,
  lightseagreen: 2142890,
  lightskyblue: 8900346,
  lightslategray: 7833753,
  lightslategrey: 7833753,
  lightsteelblue: 11584734,
  lightyellow: 16777184,
  lime: 65280,
  limegreen: 3329330,
  linen: 16445670,
  magenta: 16711935,
  maroon: 8388608,
  mediumaquamarine: 6737322,
  mediumblue: 205,
  mediumorchid: 12211667,
  mediumpurple: 9662683,
  mediumseagreen: 3978097,
  mediumslateblue: 8087790,
  mediumspringgreen: 64154,
  mediumturquoise: 4772300,
  mediumvioletred: 13047173,
  midnightblue: 1644912,
  mintcream: 16121850,
  mistyrose: 16770273,
  moccasin: 16770229,
  navajowhite: 16768685,
  navy: 128,
  oldlace: 16643558,
  olive: 8421376,
  olivedrab: 7048739,
  orange: 16753920,
  orangered: 16729344,
  orchid: 14315734,
  palegoldenrod: 15657130,
  palegreen: 10025880,
  paleturquoise: 11529966,
  palevioletred: 14381203,
  papayawhip: 16773077,
  peachpuff: 16767673,
  peru: 13468991,
  pink: 16761035,
  plum: 14524637,
  powderblue: 11591910,
  purple: 8388736,
  rebeccapurple: 6697881,
  red: 16711680,
  rosybrown: 12357519,
  royalblue: 4286945,
  saddlebrown: 9127187,
  salmon: 16416882,
  sandybrown: 16032864,
  seagreen: 3050327,
  seashell: 16774638,
  sienna: 10506797,
  silver: 12632256,
  skyblue: 8900331,
  slateblue: 6970061,
  slategray: 7372944,
  slategrey: 7372944,
  snow: 16775930,
  springgreen: 65407,
  steelblue: 4620980,
  tan: 13808780,
  teal: 32896,
  thistle: 14204888,
  tomato: 16737095,
  turquoise: 4251856,
  violet: 15631086,
  wheat: 16113331,
  white: 16777215,
  whitesmoke: 16119285,
  yellow: 16776960,
  yellowgreen: 10145074
};
sn(xt, rt, {
  copy(t) {
    return Object.assign(new this.constructor(), this, t);
  },
  displayable() {
    return this.rgb().displayable();
  },
  hex: xn,
  formatHex: xn,
  formatHex8: ei,
  formatHsl: ri,
  formatRgb: wn,
  toString: wn
});
function xn() {
  return this.rgb().formatHex();
}
function ei() {
  return this.rgb().formatHex8();
}
function ri() {
  return Gn(this).formatHsl();
}
function wn() {
  return this.rgb().formatRgb();
}
function rt(t) {
  var n, e;
  return t = (t + "").trim().toLowerCase(), (n = Kr.exec(t)) ? (e = n[1].length, n = parseInt(n[1], 16), e === 6 ? An(n) : e === 3 ? new R(n >> 8 & 15 | n >> 4 & 240, n >> 4 & 15 | n & 240, (n & 15) << 4 | n & 15, 1) : e === 8 ? wt(n >> 24 & 255, n >> 16 & 255, n >> 8 & 255, (n & 255) / 255) : e === 4 ? wt(n >> 12 & 15 | n >> 8 & 240, n >> 8 & 15 | n >> 4 & 240, n >> 4 & 15 | n & 240, ((n & 15) << 4 | n & 15) / 255) : null) : (n = Zr.exec(t)) ? new R(n[1], n[2], n[3], 1) : (n = Qr.exec(t)) ? new R(n[1] * 255 / 100, n[2] * 255 / 100, n[3] * 255 / 100, 1) : (n = Jr.exec(t)) ? wt(n[1], n[2], n[3], n[4]) : (n = jr.exec(t)) ? wt(n[1] * 255 / 100, n[2] * 255 / 100, n[3] * 255 / 100, n[4]) : (n = ti.exec(t)) ? Nn(n[1], n[2] / 100, n[3] / 100, 1) : (n = ni.exec(t)) ? Nn(n[1], n[2] / 100, n[3] / 100, n[4]) : mn.hasOwnProperty(t) ? An(mn[t]) : t === "transparent" ? new R(NaN, NaN, NaN, 0) : null;
}
function An(t) {
  return new R(t >> 16 & 255, t >> 8 & 255, t & 255, 1);
}
function wt(t, n, e, r) {
  return r <= 0 && (t = n = e = NaN), new R(t, n, e, r);
}
function ii(t) {
  return t instanceof xt || (t = rt(t)), t ? (t = t.rgb(), new R(t.r, t.g, t.b, t.opacity)) : new R();
}
function Kt(t, n, e, r) {
  return arguments.length === 1 ? ii(t) : new R(t, n, e, r ?? 1);
}
function R(t, n, e, r) {
  this.r = +t, this.g = +n, this.b = +e, this.opacity = +r;
}
sn(R, Kt, Un(xt, {
  brighter(t) {
    return t = t == null ? Tt : Math.pow(Tt, t), new R(this.r * t, this.g * t, this.b * t, this.opacity);
  },
  darker(t) {
    return t = t == null ? gt : Math.pow(gt, t), new R(this.r * t, this.g * t, this.b * t, this.opacity);
  },
  rgb() {
    return this;
  },
  clamp() {
    return new R(et(this.r), et(this.g), et(this.b), Ct(this.opacity));
  },
  displayable() {
    return -0.5 <= this.r && this.r < 255.5 && -0.5 <= this.g && this.g < 255.5 && -0.5 <= this.b && this.b < 255.5 && 0 <= this.opacity && this.opacity <= 1;
  },
  hex: bn,
  formatHex: bn,
  formatHex8: ai,
  formatRgb: $n,
  toString: $n
}));
function bn() {
  return `#${nt(this.r)}${nt(this.g)}${nt(this.b)}`;
}
function ai() {
  return `#${nt(this.r)}${nt(this.g)}${nt(this.b)}${nt((isNaN(this.opacity) ? 1 : this.opacity) * 255)}`;
}
function $n() {
  const t = Ct(this.opacity);
  return `${t === 1 ? "rgb(" : "rgba("}${et(this.r)}, ${et(this.g)}, ${et(this.b)}${t === 1 ? ")" : `, ${t})`}`;
}
function Ct(t) {
  return isNaN(t) ? 1 : Math.max(0, Math.min(1, t));
}
function et(t) {
  return Math.max(0, Math.min(255, Math.round(t) || 0));
}
function nt(t) {
  return t = et(t), (t < 16 ? "0" : "") + t.toString(16);
}
function Nn(t, n, e, r) {
  return r <= 0 ? t = n = e = NaN : e <= 0 || e >= 1 ? t = n = NaN : n <= 0 && (t = NaN), new q(t, n, e, r);
}
function Gn(t) {
  if (t instanceof q)
    return new q(t.h, t.s, t.l, t.opacity);
  if (t instanceof xt || (t = rt(t)), !t)
    return new q();
  if (t instanceof q)
    return t;
  t = t.rgb();
  var n = t.r / 255, e = t.g / 255, r = t.b / 255, i = Math.min(n, e, r), o = Math.max(n, e, r), a = NaN, s = o - i, u = (o + i) / 2;
  return s ? (n === o ? a = (e - r) / s + (e < r) * 6 : e === o ? a = (r - n) / s + 2 : a = (n - e) / s + 4, s /= u < 0.5 ? o + i : 2 - o - i, a *= 60) : s = u > 0 && u < 1 ? 0 : a, new q(a, s, u, t.opacity);
}
function oi(t, n, e, r) {
  return arguments.length === 1 ? Gn(t) : new q(t, n, e, r ?? 1);
}
function q(t, n, e, r) {
  this.h = +t, this.s = +n, this.l = +e, this.opacity = +r;
}
sn(q, oi, Un(xt, {
  brighter(t) {
    return t = t == null ? Tt : Math.pow(Tt, t), new q(this.h, this.s, this.l * t, this.opacity);
  },
  darker(t) {
    return t = t == null ? gt : Math.pow(gt, t), new q(this.h, this.s, this.l * t, this.opacity);
  },
  rgb() {
    var t = this.h % 360 + (this.h < 0) * 360, n = isNaN(t) || isNaN(this.s) ? 0 : this.s, e = this.l, r = e + (e < 0.5 ? e : 1 - e) * n, i = 2 * e - r;
    return new R(
      Yt(t >= 240 ? t - 240 : t + 120, i, r),
      Yt(t, i, r),
      Yt(t < 120 ? t + 240 : t - 120, i, r),
      this.opacity
    );
  },
  clamp() {
    return new q(Mn(this.h), At(this.s), At(this.l), Ct(this.opacity));
  },
  displayable() {
    return (0 <= this.s && this.s <= 1 || isNaN(this.s)) && 0 <= this.l && this.l <= 1 && 0 <= this.opacity && this.opacity <= 1;
  },
  formatHsl() {
    const t = Ct(this.opacity);
    return `${t === 1 ? "hsl(" : "hsla("}${Mn(this.h)}, ${At(this.s) * 100}%, ${At(this.l) * 100}%${t === 1 ? ")" : `, ${t})`}`;
  }
}));
function Mn(t) {
  return t = (t || 0) % 360, t < 0 ? t + 360 : t;
}
function At(t) {
  return Math.max(0, Math.min(1, t || 0));
}
function Yt(t, n, e) {
  return (t < 60 ? n + (e - n) * t / 60 : t < 180 ? e : t < 240 ? n + (e - n) * (240 - t) / 60 : n) * 255;
}
const un = (t) => () => t;
function si(t, n) {
  return function(e) {
    return t + e * n;
  };
}
function ui(t, n, e) {
  return t = Math.pow(t, e), n = Math.pow(n, e) - t, e = 1 / e, function(r) {
    return Math.pow(t + r * n, e);
  };
}
function li(t) {
  return (t = +t) == 1 ? Kn : function(n, e) {
    return e - n ? ui(n, e, t) : un(isNaN(n) ? e : n);
  };
}
function Kn(t, n) {
  var e = n - t;
  return e ? si(t, e) : un(isNaN(t) ? n : t);
}
const Rt = function t(n) {
  var e = li(n);
  function r(i, o) {
    var a = e((i = Kt(i)).r, (o = Kt(o)).r), s = e(i.g, o.g), u = e(i.b, o.b), l = Kn(i.opacity, o.opacity);
    return function(f) {
      return i.r = a(f), i.g = s(f), i.b = u(f), i.opacity = l(f), i + "";
    };
  }
  return r.gamma = t, r;
}(1);
function fi(t, n) {
  n || (n = []);
  var e = t ? Math.min(n.length, t.length) : 0, r = n.slice(), i;
  return function(o) {
    for (i = 0; i < e; ++i)
      r[i] = t[i] * (1 - o) + n[i] * o;
    return r;
  };
}
function ci(t) {
  return ArrayBuffer.isView(t) && !(t instanceof DataView);
}
function hi(t, n) {
  var e = n ? n.length : 0, r = t ? Math.min(e, t.length) : 0, i = new Array(r), o = new Array(e), a;
  for (a = 0; a < r; ++a)
    i[a] = It(t[a], n[a]);
  for (; a < e; ++a)
    o[a] = n[a];
  return function(s) {
    for (a = 0; a < r; ++a)
      o[a] = i[a](s);
    return o;
  };
}
function di(t, n) {
  var e = new Date();
  return t = +t, n = +n, function(r) {
    return e.setTime(t * (1 - r) + n * r), e;
  };
}
function Y(t, n) {
  return t = +t, n = +n, function(e) {
    return t * (1 - e) + n * e;
  };
}
function pi(t, n) {
  var e = {}, r = {}, i;
  (t === null || typeof t != "object") && (t = {}), (n === null || typeof n != "object") && (n = {});
  for (i in n)
    i in t ? e[i] = It(t[i], n[i]) : r[i] = n[i];
  return function(o) {
    for (i in e)
      r[i] = e[i](o);
    return r;
  };
}
var Zt = /[-+]?(?:\d+\.?\d*|\.?\d+)(?:[eE][-+]?\d+)?/g, Wt = new RegExp(Zt.source, "g");
function gi(t) {
  return function() {
    return t;
  };
}
function yi(t) {
  return function(n) {
    return t(n) + "";
  };
}
function Zn(t, n) {
  var e = Zt.lastIndex = Wt.lastIndex = 0, r, i, o, a = -1, s = [], u = [];
  for (t = t + "", n = n + ""; (r = Zt.exec(t)) && (i = Wt.exec(n)); )
    (o = i.index) > e && (o = n.slice(e, o), s[a] ? s[a] += o : s[++a] = o), (r = r[0]) === (i = i[0]) ? s[a] ? s[a] += i : s[++a] = i : (s[++a] = null, u.push({ i: a, x: Y(r, i) })), e = Wt.lastIndex;
  return e < n.length && (o = n.slice(e), s[a] ? s[a] += o : s[++a] = o), s.length < 2 ? u[0] ? yi(u[0].x) : gi(n) : (n = u.length, function(l) {
    for (var f = 0, c; f < n; ++f)
      s[(c = u[f]).i] = c.x(l);
    return s.join("");
  });
}
function It(t, n) {
  var e = typeof n, r;
  return n == null || e === "boolean" ? un(n) : (e === "number" ? Y : e === "string" ? (r = rt(n)) ? (n = r, Rt) : Zn : n instanceof rt ? Rt : n instanceof Date ? di : ci(n) ? fi : Array.isArray(n) ? hi : typeof n.valueOf != "function" && typeof n.toString != "function" || isNaN(n) ? pi : Y)(t, n);
}
var kn = 180 / Math.PI, Qt = {
  translateX: 0,
  translateY: 0,
  rotate: 0,
  skewX: 0,
  scaleX: 1,
  scaleY: 1
};
function Qn(t, n, e, r, i, o) {
  var a, s, u;
  return (a = Math.sqrt(t * t + n * n)) && (t /= a, n /= a), (u = t * e + n * r) && (e -= t * u, r -= n * u), (s = Math.sqrt(e * e + r * r)) && (e /= s, r /= s, u /= s), t * r < n * e && (t = -t, n = -n, u = -u, a = -a), {
    translateX: i,
    translateY: o,
    rotate: Math.atan2(n, t) * kn,
    skewX: Math.atan(u) * kn,
    scaleX: a,
    scaleY: s
  };
}
var bt;
function _i(t) {
  const n = new (typeof DOMMatrix == "function" ? DOMMatrix : WebKitCSSMatrix)(t + "");
  return n.isIdentity ? Qt : Qn(n.a, n.b, n.c, n.d, n.e, n.f);
}
function vi(t) {
  return t == null || (bt || (bt = document.createElementNS("http://www.w3.org/2000/svg", "g")), bt.setAttribute("transform", t), !(t = bt.transform.baseVal.consolidate())) ? Qt : (t = t.matrix, Qn(t.a, t.b, t.c, t.d, t.e, t.f));
}
function Jn(t, n, e, r) {
  function i(l) {
    return l.length ? l.pop() + " " : "";
  }
  function o(l, f, c, h, d, p) {
    if (l !== c || f !== h) {
      var y = d.push("translate(", null, n, null, e);
      p.push({ i: y - 4, x: Y(l, c) }, { i: y - 2, x: Y(f, h) });
    } else
      (c || h) && d.push("translate(" + c + n + h + e);
  }
  function a(l, f, c, h) {
    l !== f ? (l - f > 180 ? f += 360 : f - l > 180 && (l += 360), h.push({ i: c.push(i(c) + "rotate(", null, r) - 2, x: Y(l, f) })) : f && c.push(i(c) + "rotate(" + f + r);
  }
  function s(l, f, c, h) {
    l !== f ? h.push({ i: c.push(i(c) + "skewX(", null, r) - 2, x: Y(l, f) }) : f && c.push(i(c) + "skewX(" + f + r);
  }
  function u(l, f, c, h, d, p) {
    if (l !== c || f !== h) {
      var y = d.push(i(d) + "scale(", null, ",", null, ")");
      p.push({ i: y - 4, x: Y(l, c) }, { i: y - 2, x: Y(f, h) });
    } else
      (c !== 1 || h !== 1) && d.push(i(d) + "scale(" + c + "," + h + ")");
  }
  return function(l, f) {
    var c = [], h = [];
    return l = t(l), f = t(f), o(l.translateX, l.translateY, f.translateX, f.translateY, c, h), a(l.rotate, f.rotate, c, h), s(l.skewX, f.skewX, c, h), u(l.scaleX, l.scaleY, f.scaleX, f.scaleY, c, h), l = f = null, function(d) {
      for (var p = -1, y = h.length, _; ++p < y; )
        c[(_ = h[p]).i] = _.x(d);
      return c.join("");
    };
  };
}
var mi = Jn(_i, "px, ", "px)", "deg)"), xi = Jn(vi, ", ", ")", ")"), ft = 0, dt = 0, ht = 0, jn = 1e3, Ft, pt, Pt = 0, it = 0, qt = 0, _t = typeof performance == "object" && performance.now ? performance : Date, te = typeof window == "object" && window.requestAnimationFrame ? window.requestAnimationFrame.bind(window) : function(t) {
  setTimeout(t, 17);
};
function ln() {
  return it || (te(wi), it = _t.now() + qt);
}
function wi() {
  it = 0;
}
function Ot() {
  this._call = this._time = this._next = null;
}
Ot.prototype = ne.prototype = {
  constructor: Ot,
  restart: function(t, n, e) {
    if (typeof t != "function")
      throw new TypeError("callback is not a function");
    e = (e == null ? ln() : +e) + (n == null ? 0 : +n), !this._next && pt !== this && (pt ? pt._next = this : Ft = this, pt = this), this._call = t, this._time = e, Jt();
  },
  stop: function() {
    this._call && (this._call = null, this._time = 1 / 0, Jt());
  }
};
function ne(t, n, e) {
  var r = new Ot();
  return r.restart(t, n, e), r;
}
function Ai() {
  ln(), ++ft;
  for (var t = Ft, n; t; )
    (n = it - t._time) >= 0 && t._call.call(void 0, n), t = t._next;
  --ft;
}
function Sn() {
  it = (Pt = _t.now()) + qt, ft = dt = 0;
  try {
    Ai();
  } finally {
    ft = 0, $i(), it = 0;
  }
}
function bi() {
  var t = _t.now(), n = t - Pt;
  n > jn && (qt -= n, Pt = t);
}
function $i() {
  for (var t, n = Ft, e, r = 1 / 0; n; )
    n._call ? (r > n._time && (r = n._time), t = n, n = n._next) : (e = n._next, n._next = null, n = t ? t._next = e : Ft = e);
  pt = t, Jt(r);
}
function Jt(t) {
  if (!ft) {
    dt && (dt = clearTimeout(dt));
    var n = t - it;
    n > 24 ? (t < 1 / 0 && (dt = setTimeout(Sn, t - _t.now() - qt)), ht && (ht = clearInterval(ht))) : (ht || (Pt = _t.now(), ht = setInterval(bi, jn)), ft = 1, te(Sn));
  }
}
function En(t, n, e) {
  var r = new Ot();
  return n = n == null ? 0 : +n, r.restart((i) => {
    r.stop(), t(i + n);
  }, n, e), r;
}
var Ni = Fn("start", "end", "cancel", "interrupt"), Mi = [], ee = 0, Tn = 1, jt = 2, Mt = 3, Cn = 4, tn = 5, kt = 6;
function Lt(t, n, e, r, i, o) {
  var a = t.__transition;
  if (!a)
    t.__transition = {};
  else if (e in a)
    return;
  ki(t, e, {
    name: n,
    index: r,
    group: i,
    on: Ni,
    tween: Mi,
    time: o.time,
    delay: o.delay,
    duration: o.duration,
    ease: o.ease,
    timer: null,
    state: ee
  });
}
function fn(t, n) {
  var e = L(t, n);
  if (e.state > ee)
    throw new Error("too late; already scheduled");
  return e;
}
function U(t, n) {
  var e = L(t, n);
  if (e.state > Mt)
    throw new Error("too late; already running");
  return e;
}
function L(t, n) {
  var e = t.__transition;
  if (!e || !(e = e[n]))
    throw new Error("transition not found");
  return e;
}
function ki(t, n, e) {
  var r = t.__transition, i;
  r[n] = e, e.timer = ne(o, 0, e.time);
  function o(l) {
    e.state = Tn, e.timer.restart(a, e.delay, e.time), e.delay <= l && a(l - e.delay);
  }
  function a(l) {
    var f, c, h, d;
    if (e.state !== Tn)
      return u();
    for (f in r)
      if (d = r[f], d.name === e.name) {
        if (d.state === Mt)
          return En(a);
        d.state === Cn ? (d.state = kt, d.timer.stop(), d.on.call("interrupt", t, t.__data__, d.index, d.group), delete r[f]) : +f < n && (d.state = kt, d.timer.stop(), d.on.call("cancel", t, t.__data__, d.index, d.group), delete r[f]);
      }
    if (En(function() {
      e.state === Mt && (e.state = Cn, e.timer.restart(s, e.delay, e.time), s(l));
    }), e.state = jt, e.on.call("start", t, t.__data__, e.index, e.group), e.state === jt) {
      for (e.state = Mt, i = new Array(h = e.tween.length), f = 0, c = -1; f < h; ++f)
        (d = e.tween[f].value.call(t, t.__data__, e.index, e.group)) && (i[++c] = d);
      i.length = c + 1;
    }
  }
  function s(l) {
    for (var f = l < e.duration ? e.ease.call(null, l / e.duration) : (e.timer.restart(u), e.state = tn, 1), c = -1, h = i.length; ++c < h; )
      i[c].call(t, f);
    e.state === tn && (e.on.call("end", t, t.__data__, e.index, e.group), u());
  }
  function u() {
    e.state = kt, e.timer.stop(), delete r[n];
    for (var l in r)
      return;
    delete t.__transition;
  }
}
function Si(t, n) {
  var e = t.__transition, r, i, o = !0, a;
  if (!!e) {
    n = n == null ? null : n + "";
    for (a in e) {
      if ((r = e[a]).name !== n) {
        o = !1;
        continue;
      }
      i = r.state > jt && r.state < tn, r.state = kt, r.timer.stop(), r.on.call(i ? "interrupt" : "cancel", t, t.__data__, r.index, r.group), delete e[a];
    }
    o && delete t.__transition;
  }
}
function Ei(t) {
  return this.each(function() {
    Si(this, t);
  });
}
function Ti(t, n) {
  var e, r;
  return function() {
    var i = U(this, t), o = i.tween;
    if (o !== e) {
      r = e = o;
      for (var a = 0, s = r.length; a < s; ++a)
        if (r[a].name === n) {
          r = r.slice(), r.splice(a, 1);
          break;
        }
    }
    i.tween = r;
  };
}
function Ci(t, n, e) {
  var r, i;
  if (typeof e != "function")
    throw new Error();
  return function() {
    var o = U(this, t), a = o.tween;
    if (a !== r) {
      i = (r = a).slice();
      for (var s = { name: n, value: e }, u = 0, l = i.length; u < l; ++u)
        if (i[u].name === n) {
          i[u] = s;
          break;
        }
      u === l && i.push(s);
    }
    o.tween = i;
  };
}
function Ri(t, n) {
  var e = this._id;
  if (t += "", arguments.length < 2) {
    for (var r = L(this.node(), e).tween, i = 0, o = r.length, a; i < o; ++i)
      if ((a = r[i]).name === t)
        return a.value;
    return null;
  }
  return this.each((n == null ? Ti : Ci)(e, t, n));
}
function cn(t, n, e) {
  var r = t._id;
  return t.each(function() {
    var i = U(this, r);
    (i.value || (i.value = {}))[n] = e.apply(this, arguments);
  }), function(i) {
    return L(i, r).value[n];
  };
}
function re(t, n) {
  var e;
  return (typeof n == "number" ? Y : n instanceof rt ? Rt : (e = rt(n)) ? (n = e, Rt) : Zn)(t, n);
}
function Ii(t) {
  return function() {
    this.removeAttribute(t);
  };
}
function Fi(t) {
  return function() {
    this.removeAttributeNS(t.space, t.local);
  };
}
function Pi(t, n, e) {
  var r, i = e + "", o;
  return function() {
    var a = this.getAttribute(t);
    return a === i ? null : a === r ? o : o = n(r = a, e);
  };
}
function Oi(t, n, e) {
  var r, i = e + "", o;
  return function() {
    var a = this.getAttributeNS(t.space, t.local);
    return a === i ? null : a === r ? o : o = n(r = a, e);
  };
}
function Vi(t, n, e) {
  var r, i, o;
  return function() {
    var a, s = e(this), u;
    return s == null ? void this.removeAttribute(t) : (a = this.getAttribute(t), u = s + "", a === u ? null : a === r && u === i ? o : (i = u, o = n(r = a, s)));
  };
}
function Di(t, n, e) {
  var r, i, o;
  return function() {
    var a, s = e(this), u;
    return s == null ? void this.removeAttributeNS(t.space, t.local) : (a = this.getAttributeNS(t.space, t.local), u = s + "", a === u ? null : a === r && u === i ? o : (i = u, o = n(r = a, s)));
  };
}
function qi(t, n) {
  var e = Dt(t), r = e === "transform" ? xi : re;
  return this.attrTween(t, typeof n == "function" ? (e.local ? Di : Vi)(e, r, cn(this, "attr." + t, n)) : n == null ? (e.local ? Fi : Ii)(e) : (e.local ? Oi : Pi)(e, r, n));
}
function Li(t, n) {
  return function(e) {
    this.setAttribute(t, n.call(this, e));
  };
}
function Hi(t, n) {
  return function(e) {
    this.setAttributeNS(t.space, t.local, n.call(this, e));
  };
}
function Xi(t, n) {
  var e, r;
  function i() {
    var o = n.apply(this, arguments);
    return o !== r && (e = (r = o) && Hi(t, o)), e;
  }
  return i._value = n, i;
}
function zi(t, n) {
  var e, r;
  function i() {
    var o = n.apply(this, arguments);
    return o !== r && (e = (r = o) && Li(t, o)), e;
  }
  return i._value = n, i;
}
function Bi(t, n) {
  var e = "attr." + t;
  if (arguments.length < 2)
    return (e = this.tween(e)) && e._value;
  if (n == null)
    return this.tween(e, null);
  if (typeof n != "function")
    throw new Error();
  var r = Dt(t);
  return this.tween(e, (r.local ? Xi : zi)(r, n));
}
function Yi(t, n) {
  return function() {
    fn(this, t).delay = +n.apply(this, arguments);
  };
}
function Wi(t, n) {
  return n = +n, function() {
    fn(this, t).delay = n;
  };
}
function Ui(t) {
  var n = this._id;
  return arguments.length ? this.each((typeof t == "function" ? Yi : Wi)(n, t)) : L(this.node(), n).delay;
}
function Gi(t, n) {
  return function() {
    U(this, t).duration = +n.apply(this, arguments);
  };
}
function Ki(t, n) {
  return n = +n, function() {
    U(this, t).duration = n;
  };
}
function Zi(t) {
  var n = this._id;
  return arguments.length ? this.each((typeof t == "function" ? Gi : Ki)(n, t)) : L(this.node(), n).duration;
}
function Qi(t, n) {
  if (typeof n != "function")
    throw new Error();
  return function() {
    U(this, t).ease = n;
  };
}
function Ji(t) {
  var n = this._id;
  return arguments.length ? this.each(Qi(n, t)) : L(this.node(), n).ease;
}
function ji(t, n) {
  return function() {
    var e = n.apply(this, arguments);
    if (typeof e != "function")
      throw new Error();
    U(this, t).ease = e;
  };
}
function ta(t) {
  if (typeof t != "function")
    throw new Error();
  return this.each(ji(this._id, t));
}
function na(t) {
  typeof t != "function" && (t = Vn(t));
  for (var n = this._groups, e = n.length, r = new Array(e), i = 0; i < e; ++i)
    for (var o = n[i], a = o.length, s = r[i] = [], u, l = 0; l < a; ++l)
      (u = o[l]) && t.call(u, u.__data__, l, o) && s.push(u);
  return new Z(r, this._parents, this._name, this._id);
}
function ea(t) {
  if (t._id !== this._id)
    throw new Error();
  for (var n = this._groups, e = t._groups, r = n.length, i = e.length, o = Math.min(r, i), a = new Array(r), s = 0; s < o; ++s)
    for (var u = n[s], l = e[s], f = u.length, c = a[s] = new Array(f), h, d = 0; d < f; ++d)
      (h = u[d] || l[d]) && (c[d] = h);
  for (; s < r; ++s)
    a[s] = n[s];
  return new Z(a, this._parents, this._name, this._id);
}
function ra(t) {
  return (t + "").trim().split(/^|\s+/).every(function(n) {
    var e = n.indexOf(".");
    return e >= 0 && (n = n.slice(0, e)), !n || n === "start";
  });
}
function ia(t, n, e) {
  var r, i, o = ra(n) ? fn : U;
  return function() {
    var a = o(this, t), s = a.on;
    s !== r && (i = (r = s).copy()).on(n, e), a.on = i;
  };
}
function aa(t, n) {
  var e = this._id;
  return arguments.length < 2 ? L(this.node(), e).on.on(t) : this.each(ia(e, t, n));
}
function oa(t) {
  return function() {
    var n = this.parentNode;
    for (var e in this.__transition)
      if (+e !== t)
        return;
    n && n.removeChild(this);
  };
}
function sa() {
  return this.on("end.remove", oa(this._id));
}
function ua(t) {
  var n = this._name, e = this._id;
  typeof t != "function" && (t = an(t));
  for (var r = this._groups, i = r.length, o = new Array(i), a = 0; a < i; ++a)
    for (var s = r[a], u = s.length, l = o[a] = new Array(u), f, c, h = 0; h < u; ++h)
      (f = s[h]) && (c = t.call(f, f.__data__, h, s)) && ("__data__" in f && (c.__data__ = f.__data__), l[h] = c, Lt(l[h], n, e, h, l, L(f, e)));
  return new Z(o, this._parents, n, e);
}
function la(t) {
  var n = this._name, e = this._id;
  typeof t != "function" && (t = On(t));
  for (var r = this._groups, i = r.length, o = [], a = [], s = 0; s < i; ++s)
    for (var u = r[s], l = u.length, f, c = 0; c < l; ++c)
      if (f = u[c]) {
        for (var h = t.call(f, f.__data__, c, u), d, p = L(f, e), y = 0, _ = h.length; y < _; ++y)
          (d = h[y]) && Lt(d, n, e, y, h, p);
        o.push(h), a.push(f);
      }
  return new Z(o, a, n, e);
}
var fa = mt.prototype.constructor;
function ca() {
  return new fa(this._groups, this._parents);
}
function ha(t, n) {
  var e, r, i;
  return function() {
    var o = lt(this, t), a = (this.style.removeProperty(t), lt(this, t));
    return o === a ? null : o === e && a === r ? i : i = n(e = o, r = a);
  };
}
function ie(t) {
  return function() {
    this.style.removeProperty(t);
  };
}
function da(t, n, e) {
  var r, i = e + "", o;
  return function() {
    var a = lt(this, t);
    return a === i ? null : a === r ? o : o = n(r = a, e);
  };
}
function pa(t, n, e) {
  var r, i, o;
  return function() {
    var a = lt(this, t), s = e(this), u = s + "";
    return s == null && (u = s = (this.style.removeProperty(t), lt(this, t))), a === u ? null : a === r && u === i ? o : (i = u, o = n(r = a, s));
  };
}
function ga(t, n) {
  var e, r, i, o = "style." + n, a = "end." + o, s;
  return function() {
    var u = U(this, t), l = u.on, f = u.value[o] == null ? s || (s = ie(n)) : void 0;
    (l !== e || i !== f) && (r = (e = l).copy()).on(a, i = f), u.on = r;
  };
}
function ya(t, n, e) {
  var r = (t += "") == "transform" ? mi : re;
  return n == null ? this.styleTween(t, ha(t, r)).on("end.style." + t, ie(t)) : typeof n == "function" ? this.styleTween(t, pa(t, r, cn(this, "style." + t, n))).each(ga(this._id, t)) : this.styleTween(t, da(t, r, n), e).on("end.style." + t, null);
}
function _a(t, n, e) {
  return function(r) {
    this.style.setProperty(t, n.call(this, r), e);
  };
}
function va(t, n, e) {
  var r, i;
  function o() {
    var a = n.apply(this, arguments);
    return a !== i && (r = (i = a) && _a(t, a, e)), r;
  }
  return o._value = n, o;
}
function ma(t, n, e) {
  var r = "style." + (t += "");
  if (arguments.length < 2)
    return (r = this.tween(r)) && r._value;
  if (n == null)
    return this.tween(r, null);
  if (typeof n != "function")
    throw new Error();
  return this.tween(r, va(t, n, e ?? ""));
}
function xa(t) {
  return function() {
    this.textContent = t;
  };
}
function wa(t) {
  return function() {
    var n = t(this);
    this.textContent = n ?? "";
  };
}
function Aa(t) {
  return this.tween("text", typeof t == "function" ? wa(cn(this, "text", t)) : xa(t == null ? "" : t + ""));
}
function ba(t) {
  return function(n) {
    this.textContent = t.call(this, n);
  };
}
function $a(t) {
  var n, e;
  function r() {
    var i = t.apply(this, arguments);
    return i !== e && (n = (e = i) && ba(i)), n;
  }
  return r._value = t, r;
}
function Na(t) {
  var n = "text";
  if (arguments.length < 1)
    return (n = this.tween(n)) && n._value;
  if (t == null)
    return this.tween(n, null);
  if (typeof t != "function")
    throw new Error();
  return this.tween(n, $a(t));
}
function Ma() {
  for (var t = this._name, n = this._id, e = ae(), r = this._groups, i = r.length, o = 0; o < i; ++o)
    for (var a = r[o], s = a.length, u, l = 0; l < s; ++l)
      if (u = a[l]) {
        var f = L(u, n);
        Lt(u, t, e, l, a, {
          time: f.time + f.delay + f.duration,
          delay: 0,
          duration: f.duration,
          ease: f.ease
        });
      }
  return new Z(r, this._parents, t, e);
}
function ka() {
  var t, n, e = this, r = e._id, i = e.size();
  return new Promise(function(o, a) {
    var s = { value: a }, u = { value: function() {
      --i === 0 && o();
    } };
    e.each(function() {
      var l = U(this, r), f = l.on;
      f !== t && (n = (t = f).copy(), n._.cancel.push(s), n._.interrupt.push(s), n._.end.push(u)), l.on = n;
    }), i === 0 && o();
  });
}
var Sa = 0;
function Z(t, n, e, r) {
  this._groups = t, this._parents = n, this._name = e, this._id = r;
}
function ae() {
  return ++Sa;
}
var K = mt.prototype;
Z.prototype = {
  constructor: Z,
  select: ua,
  selectAll: la,
  selectChild: K.selectChild,
  selectChildren: K.selectChildren,
  filter: na,
  merge: ea,
  selection: ca,
  transition: Ma,
  call: K.call,
  nodes: K.nodes,
  node: K.node,
  size: K.size,
  empty: K.empty,
  each: K.each,
  on: aa,
  attr: qi,
  attrTween: Bi,
  style: ya,
  styleTween: ma,
  text: Aa,
  textTween: Na,
  remove: sa,
  tween: Ri,
  delay: Ui,
  duration: Zi,
  ease: Ji,
  easeVarying: ta,
  end: ka,
  [Symbol.iterator]: K[Symbol.iterator]
};
function Ea(t) {
  return ((t *= 2) <= 1 ? t * t * t : (t -= 2) * t * t + 2) / 2;
}
var hn = 1.70158;
(function t(n) {
  n = +n;
  function e(r) {
    return (r = +r) * r * (n * (r - 1) + r);
  }
  return e.overshoot = t, e;
})(hn);
var Ta = function t(n) {
  n = +n;
  function e(r) {
    return --r * r * ((r + 1) * n + r) + 1;
  }
  return e.overshoot = t, e;
}(hn);
(function t(n) {
  n = +n;
  function e(r) {
    return ((r *= 2) < 1 ? r * r * ((n + 1) * r - n) : (r -= 2) * r * ((n + 1) * r + n) + 2) / 2;
  }
  return e.overshoot = t, e;
})(hn);
var Ca = {
  time: null,
  delay: 0,
  duration: 250,
  ease: Ea
};
function Ra(t, n) {
  for (var e; !(e = t.__transition) || !(e = e[n]); )
    if (!(t = t.parentNode))
      throw new Error(`transition ${n} not found`);
  return e;
}
function Ia(t) {
  var n, e;
  t instanceof Z ? (n = t._id, t = t._name) : (n = ae(), (e = Ca).time = ln(), t = t == null ? null : t + "");
  for (var r = this._groups, i = r.length, o = 0; o < i; ++o)
    for (var a = r[o], s = a.length, u, l = 0; l < s; ++l)
      (u = a[l]) && Lt(u, t, n, l, a, e || Ra(u, n));
  return new Z(r, this._parents, t, n);
}
mt.prototype.interrupt = Ei;
mt.prototype.transition = Ia;
const nn = Math.PI, en = 2 * nn, tt = 1e-6, Fa = en - tt;
function rn() {
  this._x0 = this._y0 = this._x1 = this._y1 = null, this._ = "";
}
function oe() {
  return new rn();
}
rn.prototype = oe.prototype = {
  constructor: rn,
  moveTo: function(t, n) {
    this._ += "M" + (this._x0 = this._x1 = +t) + "," + (this._y0 = this._y1 = +n);
  },
  closePath: function() {
    this._x1 !== null && (this._x1 = this._x0, this._y1 = this._y0, this._ += "Z");
  },
  lineTo: function(t, n) {
    this._ += "L" + (this._x1 = +t) + "," + (this._y1 = +n);
  },
  quadraticCurveTo: function(t, n, e, r) {
    this._ += "Q" + +t + "," + +n + "," + (this._x1 = +e) + "," + (this._y1 = +r);
  },
  bezierCurveTo: function(t, n, e, r, i, o) {
    this._ += "C" + +t + "," + +n + "," + +e + "," + +r + "," + (this._x1 = +i) + "," + (this._y1 = +o);
  },
  arcTo: function(t, n, e, r, i) {
    t = +t, n = +n, e = +e, r = +r, i = +i;
    var o = this._x1, a = this._y1, s = e - t, u = r - n, l = o - t, f = a - n, c = l * l + f * f;
    if (i < 0)
      throw new Error("negative radius: " + i);
    if (this._x1 === null)
      this._ += "M" + (this._x1 = t) + "," + (this._y1 = n);
    else if (c > tt)
      if (!(Math.abs(f * s - u * l) > tt) || !i)
        this._ += "L" + (this._x1 = t) + "," + (this._y1 = n);
      else {
        var h = e - o, d = r - a, p = s * s + u * u, y = h * h + d * d, _ = Math.sqrt(p), N = Math.sqrt(c), w = i * Math.tan((nn - Math.acos((p + c - y) / (2 * _ * N))) / 2), x = w / N, A = w / _;
        Math.abs(x - 1) > tt && (this._ += "L" + (t + x * l) + "," + (n + x * f)), this._ += "A" + i + "," + i + ",0,0," + +(f * h > l * d) + "," + (this._x1 = t + A * s) + "," + (this._y1 = n + A * u);
      }
  },
  arc: function(t, n, e, r, i, o) {
    t = +t, n = +n, e = +e, o = !!o;
    var a = e * Math.cos(r), s = e * Math.sin(r), u = t + a, l = n + s, f = 1 ^ o, c = o ? r - i : i - r;
    if (e < 0)
      throw new Error("negative radius: " + e);
    this._x1 === null ? this._ += "M" + u + "," + l : (Math.abs(this._x1 - u) > tt || Math.abs(this._y1 - l) > tt) && (this._ += "L" + u + "," + l), e && (c < 0 && (c = c % en + en), c > Fa ? this._ += "A" + e + "," + e + ",0,1," + f + "," + (t - a) + "," + (n - s) + "A" + e + "," + e + ",0,1," + f + "," + (this._x1 = u) + "," + (this._y1 = l) : c > tt && (this._ += "A" + e + "," + e + ",0," + +(c >= nn) + "," + f + "," + (this._x1 = t + e * Math.cos(i)) + "," + (this._y1 = n + e * Math.sin(i))));
  },
  rect: function(t, n, e, r) {
    this._ += "M" + (this._x0 = this._x1 = +t) + "," + (this._y0 = this._y1 = +n) + "h" + +e + "v" + +r + "h" + -e + "Z";
  },
  toString: function() {
    return this._;
  }
};
function E(t) {
  return function() {
    return t;
  };
}
const Rn = Math.abs, S = Math.atan2, j = Math.cos, Pa = Math.max, Ut = Math.min, B = Math.sin, ot = Math.sqrt, C = 1e-12, vt = Math.PI, Vt = vt / 2, St = 2 * vt;
function Oa(t) {
  return t > 1 ? 0 : t < -1 ? vt : Math.acos(t);
}
function In(t) {
  return t >= 1 ? Vt : t <= -1 ? -Vt : Math.asin(t);
}
function Va(t) {
  return t.innerRadius;
}
function Da(t) {
  return t.outerRadius;
}
function qa(t) {
  return t.startAngle;
}
function La(t) {
  return t.endAngle;
}
function Ha(t) {
  return t && t.padAngle;
}
function Xa(t, n, e, r, i, o, a, s) {
  var u = e - t, l = r - n, f = a - i, c = s - o, h = c * u - f * l;
  if (!(h * h < C))
    return h = (f * (n - o) - c * (t - i)) / h, [t + h * u, n + h * l];
}
function $t(t, n, e, r, i, o, a) {
  var s = t - e, u = n - r, l = (a ? o : -o) / ot(s * s + u * u), f = l * u, c = -l * s, h = t + f, d = n + c, p = e + f, y = r + c, _ = (h + p) / 2, N = (d + y) / 2, w = p - h, x = y - d, A = w * w + x * x, M = i - o, k = h * y - p * d, H = (x < 0 ? -1 : 1) * ot(Pa(0, M * M * A - k * k)), X = (k * x - w * H) / A, T = (-k * w - x * H) / A, I = (k * x + w * H) / A, F = (-k * w + x * H) / A, g = X - _, m = T - N, v = I - _, b = F - N;
  return g * g + m * m > v * v + b * b && (X = I, T = F), {
    cx: X,
    cy: T,
    x01: -f,
    y01: -c,
    x11: X * (i / M - 1),
    y11: T * (i / M - 1)
  };
}
function za() {
  var t = Va, n = Da, e = E(0), r = null, i = qa, o = La, a = Ha, s = null;
  function u() {
    var l, f, c = +t.apply(this, arguments), h = +n.apply(this, arguments), d = i.apply(this, arguments) - Vt, p = o.apply(this, arguments) - Vt, y = Rn(p - d), _ = p > d;
    if (s || (s = l = oe()), h < c && (f = h, h = c, c = f), !(h > C))
      s.moveTo(0, 0);
    else if (y > St - C)
      s.moveTo(h * j(d), h * B(d)), s.arc(0, 0, h, d, p, !_), c > C && (s.moveTo(c * j(p), c * B(p)), s.arc(0, 0, c, p, d, _));
    else {
      var N = d, w = p, x = d, A = p, M = y, k = y, H = a.apply(this, arguments) / 2, X = H > C && (r ? +r.apply(this, arguments) : ot(c * c + h * h)), T = Ut(Rn(h - c) / 2, +e.apply(this, arguments)), I = T, F = T, g, m;
      if (X > C) {
        var v = In(X / c * B(H)), b = In(X / h * B(H));
        (M -= v * 2) > C ? (v *= _ ? 1 : -1, x += v, A -= v) : (M = 0, x = A = (d + p) / 2), (k -= b * 2) > C ? (b *= _ ? 1 : -1, N += b, w -= b) : (k = 0, N = w = (d + p) / 2);
      }
      var $ = h * j(N), G = h * B(N), O = c * j(A), Q = c * B(A);
      if (T > C) {
        var V = h * j(w), J = h * B(w), at = c * j(x), ct = c * B(x), D;
        if (y < vt && (D = Xa($, G, at, ct, V, J, O, Q))) {
          var Ht = $ - D[0], Xt = G - D[1], zt = V - D[0], Bt = J - D[1], dn = 1 / B(Oa((Ht * zt + Xt * Bt) / (ot(Ht * Ht + Xt * Xt) * ot(zt * zt + Bt * Bt))) / 2), pn = ot(D[0] * D[0] + D[1] * D[1]);
          I = Ut(T, (c - pn) / (dn - 1)), F = Ut(T, (h - pn) / (dn + 1));
        }
      }
      k > C ? F > C ? (g = $t(at, ct, $, G, h, F, _), m = $t(V, J, O, Q, h, F, _), s.moveTo(g.cx + g.x01, g.cy + g.y01), F < T ? s.arc(g.cx, g.cy, F, S(g.y01, g.x01), S(m.y01, m.x01), !_) : (s.arc(g.cx, g.cy, F, S(g.y01, g.x01), S(g.y11, g.x11), !_), s.arc(0, 0, h, S(g.cy + g.y11, g.cx + g.x11), S(m.cy + m.y11, m.cx + m.x11), !_), s.arc(m.cx, m.cy, F, S(m.y11, m.x11), S(m.y01, m.x01), !_))) : (s.moveTo($, G), s.arc(0, 0, h, N, w, !_)) : s.moveTo($, G), !(c > C) || !(M > C) ? s.lineTo(O, Q) : I > C ? (g = $t(O, Q, V, J, c, -I, _), m = $t($, G, at, ct, c, -I, _), s.lineTo(g.cx + g.x01, g.cy + g.y01), I < T ? s.arc(g.cx, g.cy, I, S(g.y01, g.x01), S(m.y01, m.x01), !_) : (s.arc(g.cx, g.cy, I, S(g.y01, g.x01), S(g.y11, g.x11), !_), s.arc(0, 0, c, S(g.cy + g.y11, g.cx + g.x11), S(m.cy + m.y11, m.cx + m.x11), _), s.arc(m.cx, m.cy, I, S(m.y11, m.x11), S(m.y01, m.x01), !_))) : s.arc(0, 0, c, A, x, _);
    }
    if (s.closePath(), l)
      return s = null, l + "" || null;
  }
  return u.centroid = function() {
    var l = (+t.apply(this, arguments) + +n.apply(this, arguments)) / 2, f = (+i.apply(this, arguments) + +o.apply(this, arguments)) / 2 - vt / 2;
    return [j(f) * l, B(f) * l];
  }, u.innerRadius = function(l) {
    return arguments.length ? (t = typeof l == "function" ? l : E(+l), u) : t;
  }, u.outerRadius = function(l) {
    return arguments.length ? (n = typeof l == "function" ? l : E(+l), u) : n;
  }, u.cornerRadius = function(l) {
    return arguments.length ? (e = typeof l == "function" ? l : E(+l), u) : e;
  }, u.padRadius = function(l) {
    return arguments.length ? (r = l == null ? null : typeof l == "function" ? l : E(+l), u) : r;
  }, u.startAngle = function(l) {
    return arguments.length ? (i = typeof l == "function" ? l : E(+l), u) : i;
  }, u.endAngle = function(l) {
    return arguments.length ? (o = typeof l == "function" ? l : E(+l), u) : o;
  }, u.padAngle = function(l) {
    return arguments.length ? (a = typeof l == "function" ? l : E(+l), u) : a;
  }, u.context = function(l) {
    return arguments.length ? (s = l ?? null, u) : s;
  }, u;
}
function Ba(t) {
  return typeof t == "object" && "length" in t ? t : Array.from(t);
}
function Ya(t, n) {
  return n < t ? -1 : n > t ? 1 : n >= t ? 0 : NaN;
}
function Wa(t) {
  return t;
}
function Ua() {
  var t = Wa, n = Ya, e = null, r = E(0), i = E(St), o = E(0);
  function a(s) {
    var u, l = (s = Ba(s)).length, f, c, h = 0, d = new Array(l), p = new Array(l), y = +r.apply(this, arguments), _ = Math.min(St, Math.max(-St, i.apply(this, arguments) - y)), N, w = Math.min(Math.abs(_) / l, o.apply(this, arguments)), x = w * (_ < 0 ? -1 : 1), A;
    for (u = 0; u < l; ++u)
      (A = p[d[u] = u] = +t(s[u], u, s)) > 0 && (h += A);
    for (n != null ? d.sort(function(M, k) {
      return n(p[M], p[k]);
    }) : e != null && d.sort(function(M, k) {
      return e(s[M], s[k]);
    }), u = 0, c = h ? (_ - l * x) / h : 0; u < l; ++u, y = N)
      f = d[u], A = p[f], N = y + (A > 0 ? A * c : 0) + x, p[f] = {
        data: s[f],
        index: u,
        value: A,
        startAngle: y,
        endAngle: N,
        padAngle: w
      };
    return p;
  }
  return a.value = function(s) {
    return arguments.length ? (t = typeof s == "function" ? s : E(+s), a) : t;
  }, a.sortValues = function(s) {
    return arguments.length ? (n = s, e = null, a) : n;
  }, a.sort = function(s) {
    return arguments.length ? (e = s, n = null, a) : e;
  }, a.startAngle = function(s) {
    return arguments.length ? (r = typeof s == "function" ? s : E(+s), a) : r;
  }, a.endAngle = function(s) {
    return arguments.length ? (i = typeof s == "function" ? s : E(+s), a) : i;
  }, a.padAngle = function(s) {
    return arguments.length ? (o = typeof s == "function" ? s : E(+s), a) : o;
  }, a;
}
function st(t, n, e) {
  this.k = t, this.x = n, this.y = e;
}
st.prototype = {
  constructor: st,
  scale: function(t) {
    return t === 1 ? this : new st(this.k * t, this.x, this.y);
  },
  translate: function(t, n) {
    return t === 0 & n === 0 ? this : new st(this.k, this.x + this.k * t, this.y + this.k * n);
  },
  apply: function(t) {
    return [t[0] * this.k + this.x, t[1] * this.k + this.y];
  },
  applyX: function(t) {
    return t * this.k + this.x;
  },
  applyY: function(t) {
    return t * this.k + this.y;
  },
  invert: function(t) {
    return [(t[0] - this.x) / this.k, (t[1] - this.y) / this.k];
  },
  invertX: function(t) {
    return (t - this.x) / this.k;
  },
  invertY: function(t) {
    return (t - this.y) / this.k;
  },
  rescaleX: function(t) {
    return t.copy().domain(t.range().map(this.invertX, this).map(t.invert, t));
  },
  rescaleY: function(t) {
    return t.copy().domain(t.range().map(this.invertY, this).map(t.invert, t));
  },
  toString: function() {
    return "translate(" + this.x + "," + this.y + ") scale(" + this.k + ")";
  }
};
new st(1, 0, 0);
st.prototype;
const Ka = /* @__PURE__ */ se({
  __name: "Wheel",
  props: {
    animDuration: {
      type: Number,
      default: 6e3
    },
    modelValue: {
      type: Object,
      default: () => ({}),
      validator: (t) => t.length <= 8
    },
    gift: {
      type: Number,
      default: 0
    },
    imgParams: {
      type: Object,
      default: () => ({})
    }
  },
  emits: ["done"],
  setup(t, { expose: n, emit: e }) {
    const r = t, i = z(null), o = z(!1), a = z(), s = z(null), u = z(0), l = z(0), f = z({
      width: 600,
      height: 600
    }), c = z(20), h = z(), d = z(null), p = gn(() => ({
      width: `${y.value.width}px`,
      height: `${y.value.height}px`
    })), y = gn(() => {
      const v = window.innerWidth, b = Math.min(v, f.value.width) - c.value, $ = Math.min(v, f.value.width) + 120;
      return {
        width: b,
        height: $
      };
    });
    let _ = It(0, 0);
    const N = () => {
      h.value = vn("#wheel").append("svg").attr("font-size", "16px").attr("height", "100%").attr("width", "100%").attr(
        "viewBox",
        `0 0 ${y.value.width + c.value * 2} ${y.value.height}`
      ).data([r.modelValue]).append("g").attr("class", "wrapper").attr(
        "transform",
        `translate(${(y.value.width + c.value * 2) / 2}, ${y.value.height / 2})`
      );
    }, w = () => {
      const v = h.value.append("defs").append("filter").attr("id", "shadow").attr("x", "-100%").attr("y", "-100%").attr("width", "550%").attr("height", "550%");
      v.append("feOffset").attr("in", "SourceAlpha").attr("dx", 0).attr("dy", 0).attr("result", "offsetOut"), v.append("feGaussianBlur").attr("stdDeviation", "9").attr("in", "offsetOut").attr("result", "drop"), v.append("feColorMatrix").attr("in", "drop").attr("result", "color-out").attr("type", "matrix").attr(
        "values",
        `0 0 0 0   0
          0 0 0 0   0
          0 0 0 0   0
          0 0 0 .3 0
        `
      ), v.append("feBlend").attr("in", "SourceGraphic").attr("in2", "color-out").attr("mode", "normal");
    }, x = () => {
      a.value = h.value.append("g").attr("class", "wheelholder"), d.value = a.value.append("g"), s.value = Ua().value(() => 1).padAngle(0.01).sort(null);
    }, A = () => {
      const v = za().outerRadius(u.value).innerRadius(0);
      d.value.selectAll(".middleArcText").data(s.value(r.modelValue)).enter().append("path").attr("class", "slice").attr("d", v).attr("stroke", "#000000").attr("stroke-width", "3").attr("fill", (b, $) => b.data.bgColor).each(function(b, $) {
        let O = /(^.+?)L/.exec(vn(this).attr("d"))[1].replace(/,/g, " ");
        if (b.endAngle > 90 * Math.PI / 180) {
          const Q = /M(.*?)A/, V = /A(.*?)0 0 1/, at = /0 0 1 (.*?)$/.exec(O)[1], ct = Q.exec(O)[1], D = V.exec(O)[1];
          O = `M${at}A${D}0 0 0${ct}`;
        }
        d.value.append("path").attr("class", "hiddenarcs").attr("id", "middleArc" + $).attr("d", O).style("fill", "none");
      });
    }, M = () => {
      d.value.selectAll(".middleArcText").data(s.value(r.modelValue)).enter().append("text").attr("class", "middleArcText").attr("dy", (v) => v.endAngle > 90 * Math.PI / 180 ? -35 : 42).append("textPath").attr("startOffset", "50%").attr("text-anchor", "middle").attr("fill", (v, b) => v.data.value.color).attr("xlink:href", (v, b) => "#middleArc" + b).text((v) => v.data.value);
    }, k = () => {
      a.value.append("circle").attr("cx", 0).attr("cy", 0).attr("r", u.value / 2.5).attr("fill", "#ffffff").attr("filter", "url(#shadow)").attr("stroke-width", 4).attr("stroke", "#000000");
    }, H = () => {
      a.value.append("g").append("circle").attr("cx", 0).attr("cy", 0).attr("r", (y.value.width - 4) / 2).attr("fill", "transparent").attr("stroke-width", "8").attr("filter", "url(#shadow)").attr("stroke", "#ffffff");
    }, X = () => {
      const { width: v, height: b, src: $ } = r.imgParams;
      a.value.append("svg:image").attr("x", `-${v / 2}`).attr("y", `-${v / 2}`).attr("width", v).attr("height", b).attr("xlink:href", $);
    }, T = () => {
      const v = "M95.3,9.8c-16.5,0-23.7,15.6-21.9,27c3.4,21.7,21.9,42.2,21.9,42.2s18.5-20.5,21.9-42.2 C118.9,25.4,111.8,9.8,95.3,9.8z";
      a.value && (i.value = a.value.append("g").append("path").attr("d", v).attr("transform", "translate(-20, -300)").attr("stroke", "#ffffff").attr("fill", "#FFFFFF").attr("filter", "url(#shadow)").attr(
        "transform",
        `matrix(1, 0, 0, 1, -95, -${y.value.height / 2 - c.value})`
      ).attr("stroke-linejoin", "round").attr("stroke-width", "4"));
    }, I = (v) => r.modelValue.findIndex((b) => b.id === v) + 1, F = () => (v) => `rotate(${_(v)})`, g = () => {
      N(), w(), x(), A(), M(), k(), H(), r.imgParams && X(), T();
    };
    return n({
      spin: async () => {
        if (!o.value) {
          o.value = !0;
          const v = I(r.gift), b = r.modelValue.length, $ = 360 / b, G = 360 - $ * (v - 1), O = 360 * 5, Q = G + O;
          l.value = Math.round(Q / $) * $;
          let V = Math.round(b - l.value % 360 / $);
          V = V >= b ? V % b : V;
          const J = $ + $ / 2;
          l.value += J - Math.round($ * 2), _ = It(0, l.value), await (() => d.value.transition().duration(r.animDuration).ease(Ta.overshoot(0.3)).attrTween("transform", F).end())(), o.value = !1, e("done", r.modelValue[V]);
        }
      }
    }), ue(() => {
      u.value = Math.min(y.value.width, y.value.height) / 2, g();
    }), (v, b) => (le(), fe("div", {
      id: "wheel",
      class: ce(["wheel", `wheel_font-size--${t.modelValue.length}`]),
      style: he(de(p))
    }, null, 6));
  }
});
export {
  Ka as Wheel
};
